eskip/predicates.go

```diff
@@ -0,0 +1,98 @@
+package eskip
+
+import (
+	"fmt"
+)
+
+// PredicatesContain checks if the route has predicate with the given name.
+func PredicatesContain(p []*Predicate, name string) bool {
+	for _, p := range p {
+		if p.Name == name {
+			return true
+		}
+	}
+
+	return false
+}
+
+// AllPredicatesByName returns all predicates that match the given name.
+func AllPredicatesByName(p []*Predicate, name string) []*Predicate {
+	pp := make([]*Predicate, 0)
+	for _, p := range p {
+		if p.Name == name {
+			pp = append(pp, p)
+		}
+	}
+
+	return pp
+}
+
+// SinglePredicateByName returns matching predicate or error when multiple predicates are given.
+func SinglePredicateByName(p []*Predicate, name string) (*Predicate, error) {
+	pp := make([]*Predicate, 0)
+	for _, p := range p {
+		if p.Name == name {
+			pp = append(pp, p)
+		}
+	}
+
+	if len(pp) == 0 {
+		return nil, nil
+	}
+
+	if len(pp) > 1 {
+		return nil, fmt.Errorf("multiple predicates of the same name: %d %s", len(pp), name)
+	}
+
+	return pp[0], nil
+}
+
+// ValidatePredicates returns an error when certain predicates are added multiple times.
+func ValidatePredicates(pp []*Predicate) error {
+	for _, p := range pp {
+		switch p.Name {
+		case "Path", "Weight":
```

Comment:
Can include the PathSubtree handling in this PR, too? The rules should be sg like this:

- 0 or 1 Weight predicates
- 0 or 1 of either Path or PathSubtree predicates

Reply:
Let's do this in a new PR,

Reply:
Maybe you can create an issue to track it then I'll follow up on it.